File: sdk/chocobase-java/src/main/java/io/chocobase/Storage.java
package io.chocobase;

import com.google.gson.Gson;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.HashMap;
import java.util.Map;

public class Storage {
    private final String baseUrl;
    private final HttpClient httpClient;
    private final Map<String, String> headers;
    private final Gson gson = new Gson();

    public Storage(String baseUrl, HttpClient httpClient, Map<String, String> headers) {
        this.baseUrl = baseUrl;
        this.httpClient = httpClient;
        this.headers = headers;
    }

    public StorageBucket from(String bucket) {
        return new StorageBucket(baseUrl, bucket, httpClient, headers);
    }

    public static class StorageBucket {
        private final String baseUrl;
        private final String bucket;
        private final HttpClient httpClient;
        private final Map<String, String> headers;
        private final Gson gson = new Gson();

        public StorageBucket(String baseUrl, String bucket, HttpClient httpClient, Map<String, String> headers) {
            this.baseUrl = baseUrl;
            this.bucket = bucket;
            this.httpClient = httpClient;
            this.headers = headers;
        }

        public String createSignedUrl(String path, int expiresIn) throws Exception {
            String url = baseUrl + "/v1/storage/v1/object/sign/" + bucket + "/" + path;
            Map<String, Integer> payload = new HashMap<>();
            payload.put("expires_in", expiresIn);

            HttpRequest.Builder builder = HttpRequest.newBuilder()
                    .uri(URI.create(url))
                    .header("Content-Type", "application/json")
                    .POST(HttpRequest.BodyPublishers.ofString(gson.toJson(payload)));

            headers.forEach(builder::header);
            HttpResponse<String> resp = httpClient.send(builder.build(), HttpResponse.BodyHandlers.ofString());
            Map<String, Object> map = gson.fromJson(resp.body(), Map.class);
            if (map != null && map.containsKey("signed_url")) {
                return baseUrl + map.get("signed_url");
            }
            return null;
        }
    }
}

File: sdk/chocobase-kt/src/main/kotlin/io/chocobase/Storage.kt
package io.chocobase

import com.google.gson.Gson
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody

class StorageBucket(
    private val baseUrl: String,
    private val bucket: String,
    private val httpClient: OkHttpClient,
    private val headers: Map<String, String>
) {
    private val gson = Gson()
    private val jsonType = "application/json; charset=utf-8".toMediaType()

    fun createSignedUrl(path: String, expiresIn: Int = 3600): String? {
        val payload = gson.toJson(mapOf("expires_in" to expiresIn))
        val builder = Request.Builder()
            .url("$baseUrl/v1/storage/v1/object/sign/$bucket/$path")
            .post(payload.toRequestBody(jsonType))

        headers.forEach { (k, v) -> builder.addHeader(k, v) }

        httpClient.newCall(builder.build()).execute().use { response ->
            if (response.isSuccessful) {
                val body = response.body?.string() ?: return null
                val map = gson.fromJson(body, Map::class.java)
                val signedUrl = map["signed_url"] as? String ?: return null
                return "$baseUrl$signedUrl"
            }
            return null
        }
    }
}

class StorageClient(
    private val baseUrl: String,
    private val httpClient: OkHttpClient,
    private val headers: Map<String, String>
) {
    fun from(bucket: String): StorageBucket {
        return StorageBucket(baseUrl, bucket, httpClient, headers)
    }
}
